domain/port/service/service.go

```diff
@@ -76,6 +82,57 @@ func (s *Service) GetMachineOpenedPorts(ctx context.Context, machineUUID string)
 	return s.st.GetMachineOpenedPorts(ctx, machineUUID)
 }
 
+// GetApplicationOpenedPorts returns the opened ports for all the units of the
+// application. Opened ports are grouped first by unit and then by endpoint.
+func (s *Service) GetApplicationOpenedPorts(ctx context.Context, applicationUUID string) (map[string]network.GroupedPortRanges, error) {
+	return s.st.GetApplicationOpenedPorts(ctx, applicationUUID)
+}
+
+// GetApplicationOpenedPortsByEndpoint returns all the opened ports for the given
+// application, across all units, grouped by endpoint.
+//
+// NOTE: The returned port ranges are atomised, meaning that each port range
+// we guarantee that each port range is of unit length. This is useful for
+// down-stream consumers such as k8s, which can only reason with unit-length
+// port ranges.
+func (s *Service) GetApplicationOpenedPortsByEndpoint(ctx context.Context, applicationUUID string) (network.GroupedPortRanges, error) {
+	openedPorts, err := s.st.GetApplicationOpenedPorts(ctx, applicationUUID)
+	if err != nil {
+		return nil, err
+	}
+	ret := network.GroupedPortRanges{}
+
+	// combine port ranges across all units & atomise the port ranges
+	for _, grp := range openedPorts {
```

Comment:
I don't mind too much because none of it is being done inside a transaction, but there's a redundancy whereby:
- You group by unit in state via `groupPortRangesByUnitByEndpoint`.
- You then flatten out the unit grouping.

If you had it flat from state, then you could do a single loop ever each endpoint to atomise and de-dupe.

Take it as a nit.

Reply:
Refactored